sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/state/RocksDBFileManager.scala

```diff
@@ -0,0 +1,165 @@
+
+package org.apache.spark.sql.execution.streaming.state
+
+import java.io.File
+import java.nio.charset.StandardCharsets.UTF_8
+import java.nio.file.Files
+
+import scala.collection.Seq
+
+import com.fasterxml.jackson.annotation.JsonInclude.Include
+import com.fasterxml.jackson.databind.{DeserializationFeature, ObjectMapper}
+import com.fasterxml.jackson.module.scala.{DefaultScalaModule, ScalaObjectMapper}
+import org.json4s.NoTypeHints
+import org.json4s.jackson.Serialization
+
+/**
+ * Classes to represent metadata of checkpoints saved to DFS. Since this is converted to JSON, any
+ * changes to this MUST be backward-compatible.
+ */
+case class RocksDBCheckpointMetadata(
+    sstFiles: Seq[RocksDBSstFile],
+    logFiles: Seq[RocksDBLogFile],
+    numKeys: Long) {
+  import RocksDBCheckpointMetadata._
+
+  def json: String = {
+    // We turn this field into a null to avoid write a empty logFiles field in the json.
+    val nullified = if (logFiles.isEmpty) this.copy(logFiles = null) else this
+    mapper.writeValueAsString(nullified)
+  }
+
+  def prettyJson: String = Serialization.writePretty(this)(RocksDBCheckpointMetadata.format)
+
+  def writeToFile(metadataFile: File): Unit = {
+    val writer = Files.newBufferedWriter(metadataFile.toPath, UTF_8)
+    try {
+      writer.write(s"v$VERSION\n")
+      writer.write(this.json)
+    } finally {
+      writer.close()
+    }
+  }
+
+  def immutableFiles: Seq[RocksDBImmutableFile] = sstFiles ++ logFiles
+}
+
+/** Helper class for [[RocksDBCheckpointMetadata]] */
+object RocksDBCheckpointMetadata {
+  val VERSION = 1
+
+  implicit val format = Serialization.formats(NoTypeHints)
+
+  /** Used to convert between classes and JSON. */
+  lazy val mapper = {
+    val _mapper = new ObjectMapper with ScalaObjectMapper
+    _mapper.setSerializationInclusion(Include.NON_ABSENT)
+    _mapper.configure(DeserializationFeature.FAIL_ON_UNKNOWN_PROPERTIES, false)
+    _mapper.registerModule(DefaultScalaModule)
+    _mapper
+  }
+
+  def readFromFile(metadataFile: File): RocksDBCheckpointMetadata = {
+    val reader = Files.newBufferedReader(metadataFile.toPath, UTF_8)
+    try {
+      val versionLine = reader.readLine()
+      if (versionLine != s"v$VERSION") {
+        throw new IllegalStateException(
+          s"Cannot read RocksDB checkpoint metadata of version $versionLine")
+      }
+      Serialization.read[RocksDBCheckpointMetadata](reader)
+    } finally {
+      reader.close()
+    }
+  }
+
+  def apply(rocksDBFiles: Seq[RocksDBImmutableFile], numKeys: Long): RocksDBCheckpointMetadata = {
+    val sstFiles = rocksDBFiles.collect { case file: RocksDBSstFile => file }
+    val logFiles = rocksDBFiles.collect { case file: RocksDBLogFile => file }
+
+    RocksDBCheckpointMetadata(sstFiles, logFiles, numKeys)
+  }
+}
+
+/**
+ * A RocksDBImmutableFile maintains a mapping between a local RocksDB file name and the name of
+ * its copy on DFS. Since these files are immutable, their DFS copies can be reused.
```

Comment:
Does it mean that a DFS copy can be mapped to more than one local file names?

Reply:
When do we reuse the DFS copies?

Reply:
Yes. Can be mapped to more than one local file but for different tasks. The most common scenario is task/stage retry.